invisible_cities/reco/psf_functions.py

```diff
@@ -0,0 +1,133 @@
+    bin_centers : Bin centers of the PSF.
+    """
+
+    entries, edges = np.histogramdd(pos, nbins, range=ranges, normed=False)
+    sumC   , edges = np.histogramdd(pos, nbins, range=ranges, normed=False, weights=charge)
+    with np.errstate(divide='ignore', invalid='ignore'):
+        psf = np.nan_to_num(sumC/entries)
+
+    return psf, entries, [shift_to_bin_centers(edge) for edge in edges]
+
+
+def hdst_PSF_processing(dsts        : pd.DataFrame,
+                        ranges      : List[List[float]],
+                        sampleWidth : List[float]
+                        ) -> pd.DataFrame :
+    """
+    Adds the necessary info to a hits DST to create the PSF, namely the relative position and the normalized Q.
+
+    Parameters
+    ----------
+    dsts        : hits (1 SiPM per hit).
+    ranges      : range of the PSF in each dimension.
+    sampleWidth : Sampling distance of the hits.
+
+    Returns
+    ----------
+    hdst        : hits after processing to create PSF.
+    """
+    def AddVariableWeightedMean(df        : pd.DataFrame,
+                                varMean   : str,
+                                varWeight : str
+                                ) -> pd.DataFrame :
+        """
+        Adds the average of a variable weighted by another to a
+        grouped hits DST 'df' (grouped using groupby, by event id).
+
+        Parameters
+        ----------
+        df        : groupby by event and npeak.
+        varMean   : variable to be averaged.
+        varWeight : variable to be uses as the weight.
+
+        Returns
+        ----------
+        df        : dst with the weighted average.
+        """
+        df[varMean + 'peak'] =  (df[varMean]*df[varWeight]).sum()/df[varWeight].sum()
+        return(df)
+
+    def AddEmptySensors(df  : pd.DataFrame,
+                        var : List[str]
+                        ) -> pd.DataFrame :
+        """
+        Adds empty sensors to the hDST
+
+        Parameters
+        ----------
+        df  : groupby by event and npeak.
+        var : dimensions to be considered.
+
+        Returns
+        ----------
+        df  : dst with empty sipm hits.
+        """
+        distance = (np.diff(ranges)/2).flatten()
+        means    = [int(df[f'{v}peak'].mean()) for v in var[:len(ranges)]]
+        means    = [mean - mean%sampleWidth[i]  for i, mean in enumerate(   means)]
+        varrange = [[means[i] - d, means[i] + d + sampleWidth[i]]                   for i,    d in enumerate(distance)]
+        allbins  = [int(np.diff(rang)/sampleWidth[i])              for i, rang in enumerate(varrange)]
+        Hs, edges   = np.histogramdd(tuple(df[v].values for v in var[:len(ranges)]), bins=allbins, range=varrange, normed=False, weights=df['Q'])
+        interPoints = np.meshgrid(*(shift_to_bin_centers(edge) for edge in edges), indexing='ij')
+        if len(ranges) < 3:
+            interPoints.append(np.array([df['Z'].min()] * len(interPoints[0].flatten())))
```

Comment:
If I recall correctly:

Line 93-94: Get the mean position of the event and shift it always downwards (I've had problems to properly center everything if that changes). Knowing that the SiPM position is at  10*n + 5 mm I could have tried to be smart and move it upwards when the mean is above 10*n + 5 mm. However, since for future detectors, the pitch won't be the same, I felt inclined to just use a fixed decission which is to go downwards. But it's kind of arbitrary. I just want it to always be, relatively speaking, on the same place. In any case, it does not make a difference as long as the range, for adding empty SiPMs later, is large enough.

Line 95: an extra sampleWidth is added to one side, this is to compensate for the systematic downward choice made before. 

The rest is just plotting the SiPMs charge in a grid with bin centers equal to the sensors position, so I can later create a dataframe with the values of the histogram (filled for the sensors that have charge, 0 when the SiPM was not in the DST). This is as fast as I could think of for adding the 0s. I think I could have used pad but I would still have to put the charge SiPMs in a given grid and I think the histogram option would still be the easiest way to do so.